text/0000-permissions.md

```diff
@@ -0,0 +1,186 @@
+```rust
+impl Person {
+  pub fn aging(&mut self) {
+    self.age += 1;
+  }
+  
+  pub fn rename(&mut self, new_name: String) -> String {
+    std::mem::replace(&mut self.name, new_name)
+  }
+}
+```
+Not too fancy, right? But we now want a `Department` for changing his name and a `Calendar` to age a person by one year.
+```rust
+pub struct Department<'p>(&'p mut Person); // ideally we want more than one person, but "simplicity"
+
+impl Department<'_> {
+  pub fn rename(&self, new_name: String) -> String {
+    self.person.rename(new_name)
+  }
+}
+
+pub struct Calendar<'p>(&'p mut Person); // ideally we want more than one person, but "simplicity"
+
+impl Calendar<'_> {
+  pub fn next_year(&self) {
+    self.person.aging();
+  }
+}
+```
+Still simple, right? So let us create some life 😊
+```rust
+fn main() {
+  let mut person = Person {
+    name: "Mike",
+    age: 0,
+  };
+  
+  let department = Department(&mut person);
+  let calendar = Calendar(&mut person); // ERROR: cannot borrow `*person` as mutable more than once at a time second mutable borrow occurs here
+  
+  calendar.next_year();
+  calendar.next_year();
+  calendar.next_year();
+  department.rename("Jonathan");
+  calendar.next_year();
+}
+```
+Dang it! But with permissions, we can easily fix this. So let us change some things in our `impl` of `Person`:
+```rust
+impl Person {
+  pub fn aging(&'Age mut self) { // this function needs the permission `'Age`
+    self.age += 1;
+  }
+  
+  pub fn rename(&'Name mut self, new_name: String) -> String { // this function needs the permission `'Name`
+    std::mem::replace(&mut self.name, new_name)
+  }
+}
+```
+Now let us change the borrow for `Department` and `Calendar`:
+```rust
+pub struct Department<'p>(&'p 'Name mut Person); // the borrow has the permission `'Name`
+pub struct Calendar<'p>(&'p 'Age mut Person); // the borrow has the permission `'Age`
+```
+And now let's see our main code:
+```rust
+fn main() {
+  let mut person = Person {
+    name: "Mike",
+    age: 0,
+  };
+  
+  let department = Department(&mut person); // implicit permission
+  let calendar = Calendar(&'Age mut person); // explicit permission
+  
+  calendar.next_year();
+  calendar.next_year();
+  calendar.next_year();
+  department.rename("Jonathan");
+  calendar.next_year();
+}
+```
+And TADA: our code compiles and works 🥳.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The compiler should know what the permissions allow of data to use.
+For example, when two functions need the same data mutably and normally it wouldn't compile, they shouldn't just add two different permissions to evade it:
+```rust
+impl Hack {
+  pub fn 'HackA hack_a(&mut self) {
+    self.one_data./*...*/
+  }
+  
+  pub fn 'HackB hack_b(&mut self) {
+    self.one_data./*...*/
+  }
+}
+```
+Instead, the compiler should notify that the permissions intercept with each other and they should create a sub-permission for both:
```

Comment:
So the compiler checks which (private) fields the functions use? This seems to be very much like #1215.